include/verilated_vpi.cpp

```diff
@@ -630,6 +630,96 @@ class VerilatedVpiCbHolder final {
     void invalidate() { m_id = 0; }
 };
 
+class VerilatedVpiPutHolder final {
+    const VerilatedVar* m_varp;
+    const VerilatedScope* m_scopep;
+    s_vpi_value m_value;
+    std::string m_str;
+    std::vector<s_vpi_vecval> m_vector;
+
+public:
+    VerilatedVpiPutHolder(const VerilatedVpioVar* vop, p_vpi_value valuep)
+        : m_varp(vop->varp())
+        , m_scopep(vop->scopep()) {
+        m_value.format = valuep->format;
+        switch (valuep->format) {
+        case vpiBinStrVal:
+        case vpiOctStrVal:
+        case vpiDecStrVal:
+        case vpiHexStrVal:
+        case vpiStringVal: {
+            m_str = valuep->value.str;
+            m_value.value.str = &m_str[0];
+            break;
+        }
```

Comment:
Is this function duplicated code we should put into a subroutine?

Reply:
> Is this function duplicated code we should put into a subroutine?

I don't believe so.  It's setting some members of a class which was created for this PR, so I don't believe it could have existed beforehand.  And there's nothing else that sets `m_str`.

But maybe I'm missing your point.  If there's something else that does this can you point me at it so I can refactor?